### PR TITLE
Security: fix joining cluster with production license

### LINKED ISSUE #31332: Nodes without security explicitly enabled cannot join cluster with security enabled.
This seems to be caused by the change to made security disabled by default on trial licenses.

If a node does not set `xpack.security.enabled` in the config, and tries to join a cluster, it doesn't know what type of license its supposed to be using, and seems to be acting as if its on a trial license with security disabled.

This means that `SecurityServerTransportInterceptor.interceptSender().AsyncSender.sendRequest()`  does not intercept the request, and does not set the system-user context for outgoing requests.

Since the requests do not have an authentication header attached to their context, the remote node rejects the initial handshake, and the new node cannot join.

### PR DESCRIPTION
**Pull request by jaymode**
The changes made to disable security for trial licenses unless security
is explicitly enabled caused issues when a 6.3 node attempts to join a
cluster that already has a production license installed. The new node
starts off with a trial licenses and `xpack.security.enabled` is not
set for the node, which causes the security code to skip attaching the
user to the request. The existing cluster has security enabled and the
lack of a user attached to the requests causes the request to be
rejected.

This commit changes the security code to check if the state has been
recovered yet when making the decision on whether or not to attach a
user. If the state has not yet been recovered, the code will attach
the user to the request in case security is enabled on the cluster
being joined.

Closes #31332

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java [hunk 1]
```diff
@@ -254,7 +254,8 @@
 
     public XPackLicenseState(Settings settings) {
         this.isSecurityEnabled = XPackSettings.SECURITY_ENABLED.get(settings);
-        this.isSecurityExplicitlyEnabled = settings.hasValue(XPackSettings.SECURITY_ENABLED.getKey()) && isSecurityEnabled;
+        this.isSecurityExplicitlyEnabled = (isSecurityEnabled && settings.hasValue(XPackSettings.SECURITY_ENABLED.getKey())) ||
```
Comment (jaymode) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java]: this was a suggestion from @bleskes that I went ahead and adopted as well

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java [hunk 2]
```diff
@@ -98,7 +104,9 @@ public AsyncSender interceptSender(AsyncSender sender) {
             @Override
             public <T extends TransportResponse> void sendRequest(Transport.Connection connection, String action, TransportRequest request,
                                                                   TransportRequestOptions options, TransportResponseHandler<T> handler) {
-                if (licenseState.isSecurityEnabled() && licenseState.isAuthAllowed()) {
+                final boolean stateNotRecovered = isStateNotRecovered;
```
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java]: The purposes of this assignment (from volatile to local) initially confused me, and I'm worried someone might think it is redundant and remove it in the future. Can we add a comment to try and discourage that from happening?

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java [hunk 3]
```diff
@@ -98,7 +104,9 @@ public AsyncSender interceptSender(AsyncSender sender) {
...
+                final boolean sendWithAuth = (licenseState.isSecurityEnabled() && licenseState.isAuthAllowed()) || stateNotRecovered;
```
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java]: It seems that if the state is not recovered, then we're going to `sendWithAuth` even if the local node explitily disables security.
If so, then why have the check at all? If it's safe to send the auth context it even if security is disabled, isn't it simpler to just send it for every outgoing request?

**Review by tvernum — COMMENTED** [review]
I have a question about the choice to always send auth if the state is not recovered (irrespective of the node settings)

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java (hunk 3 above)
Comment (ywelsch) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java]: Same thought as @tvernum here.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java (hunk 1 above)
Comment (jaymode) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java]: This is for cases where a user with a production license on 6.0 to 6.2.x upgrades to 6.3.1+. We require TLS for production mode, so if TLS is enabled we can take that as security being explicitly enabled.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java (hunk 3 above)
Comment (jaymode) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java]: The interceptor is not registered at all if security is disabled.

https://github.com/elastic/elasticsearch/blob/3b70e943eb08f90aa24e9c34033c8ca2a7211239/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java#L808-L829

This change limits to only sending auth when security *could* possibly be enabled. Once we've recovered the state then the check is based upon the license. By having the check we maintain the ability to assert that all outgoing requests have authentication after the state has been recovered.

**Review by tvernum — APPROVED** [review]
LGTM

**Review by ywelsch — APPROVED** [review]
LGTM

**Pull request merged**